src/jws/flattened.rs

```diff
@@ -0,0 +1,542 @@
+        let protected_header_serialized = serialize_header(&header);
+        let protected_header_registered = header.registered;
+        Self {
+            protected_header_registered,
+            protected_header_serialized,
+            payload,
+        }
+    }
+
+    /// JWS Signing Input
+    fn signing_input(&self) -> Vec<u8> {
+        signing_input(&self.protected_header_serialized, &self.payload)
+    }
+
+    /// Return a reference to the registered (known to biscuit)
+    /// protected headers
+    pub fn protected_header_registered(&self) -> &RegisteredHeader {
+        &self.protected_header_registered
+    }
+
+    /// Return a reference to protected headers as they were serialized
+    pub fn protected_header_serialized(&self) -> &[u8] {
+        &self.protected_header_serialized
+    }
+
+    /// Deserialize protected headers
+    /// This allows access to protected headers beyond those
+    /// that are recognized with RegisteredHeader
+    pub fn deserialize_protected_header<H: DeserializeOwned>(
+        &self,
+    ) -> serde_json::Result<Header<H>> {
+        serde_json::from_slice(&self.protected_header_serialized)
+    }
+
+    /// Return a reference to the payload bytes
+    pub fn payload(&self) -> &[u8] {
+        &self.payload
+    }
+
+    /// Deserialize a JSON payload
+    /// NOTE: JWS does not put any requirement on payload bytes, which
+    /// need not be JSON
+    pub fn deserialize_json_payload<T: DeserializeOwned>(&self) -> serde_json::Result<T> {
+        serde_json::from_slice(&self.payload)
+    }
+}
+
+/// Signed data (with a single signature)
+///
+/// This representation preserves the exact serialisation of
+/// the payload and protected headers, but it is independent of how
+/// the signature may be serialized (eg, flattened or compact JWS)
+///
+/// Signed data can be obtained by either deserializing a valid JWS,
+/// or by signing a Signable
+#[derive(Clone)]
+pub struct SignedData {
+    data: Signable,
+    secret: Secret,
+    signature: Vec<u8>,
+}
+
+impl SignedData {
+    /// Sign using a secret
+    pub fn sign(data: Signable, secret: Secret) -> Result<Self, Error> {
+        let signature = data
+            .protected_header_registered
+            .algorithm
+            .sign(&data.signing_input(), &secret)?;
+        Ok(Self {
+            data,
+            secret,
+            signature,
+        })
+    }
+
+    /// Serialize using Flattened JWS JSON Serialization
+    /// See RFC 7515 section 7.2.2
+    pub fn serialize_flattened(&self) -> String {
+        let payload = self.data.payload.clone();
+        let protected_header = self.data.protected_header_serialized.clone();
+        let signature = self.signature.clone();
+        let s = FlattenedRaw {
+            payload,
+            protected_header,
+            signature,
+            signatures: (),
+            unprotected_header: (),
+        };
+        serde_json::to_string(&s).unwrap()
```

Comment:
Let's avoid unwrapping and return `Result`.

Reply:
I added a comment explaining why this shouldn't fail instead.